### PR TITLE
Implement the Basic HTTP package provider

### PR DESCRIPTION
**Pull request by abul-wcse**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -3,6 +3,7 @@
     "type": "library",
     "require": {
         "php": ">=5.6",
+        "guzzlehttp/guzzle": "~6.0",
```
Comment (jedkirby) [inline composer.json:6]: Is this going to cause any issues with customers which are not using this version of Guzzle, or, using the _really_ old version, i.e. `guzzle/guzzle`?

File: src/Package/Provider/BasicHttp/BasicHttpPackageProvider.php [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+namespace Meteor\Package\Provider\BasicHttp;
+
+use GuzzleHttp\Client;
+use Meteor\IO\IOInterface;
+use Meteor\Package\Provider\Exception\PackageNotFoundException;
+use Meteor\Package\Provider\PackageProviderInterface;
+
+class BasicHttpPackageProvider implements PackageProviderInterface
+{
+
```
Comment (jedkirby) [inline src/Package/Provider/BasicHttp/BasicHttpPackageProvider.php]: Think PHPCS needs to be run, this space shouldn't be present.

File: src/Package/Provider/BasicHttp/BasicHttpPackageProvider.php [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        $file = $version . '.zip';
+        $baseUrl = $this->packageBaseUrls[$packageName];
+
+        $packageURL = $baseUrl . $file;
+        try {
+            $response = $this->httpClient->request('GET', $packageURL, ['sink' => $tempDir . '/' . $file]);
+            if ($response->getStatusCode() == 200 && file_exists($tempDir . '/' . $file)) {
+                return $tempDir . '/' . $file;
+            }
+        } catch (\Exception $e) {
```
Comment (jedkirby) [inline src/Package/Provider/BasicHttp/BasicHttpPackageProvider.php]: Should we add any logging here, or, maybe output to help diagnose the issue, i.e. `$e->getMessage()`?

File: src/Package/Provider/BasicHttp/BasicHttpPackageProvider.php [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+     */
+    public function download($packageName, $version, $tempDir)
+    {
+        if (!isset($this->packageBaseUrls[$packageName])) {
+            throw new PackageNotFoundException(sprintf('The HTTP base url for the "%s" package has not been configured.', $packageName));
+        }
+        $file = $version . '.zip';
+        $baseUrl = $this->packageBaseUrls[$packageName];
+
+        $packageURL = $baseUrl . $file;
```
Comment (jedkirby) [inline src/Package/Provider/BasicHttp/BasicHttpPackageProvider.php]: I'd argue that `$packageURL` should be `$packageUrl`, for consistency with `$this->packageBaseUrls`.

File: src/Package/Provider/BasicHttp/BasicHttpPackageProvider.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        try {
+            $response = $this->httpClient->request('GET', $packageURL, ['sink' => $tempDir . '/' . $file]);
+            if ($response->getStatusCode() == 200 && file_exists($tempDir . '/' . $file)) {
+                return $tempDir . '/' . $file;
```
Comment (jedkirby) [inline src/Package/Provider/BasicHttp/BasicHttpPackageProvider.php]: Do we need to remove the file after downloading it?

File: src/Package/Provider/BasicHttp/BasicHttpPackageProvider.php (hunk 3 above)
Comment (jedkirby) [reply src/Package/Provider/BasicHttp/BasicHttpPackageProvider.php]: Also, import the exception class.

**Review by jedkirby — CHANGES_REQUESTED** [review]
Generally looks good, few comments but nothing too serious.

File: composer.json (hunk 1 above)
Comment (jedkirby) [reply composer.json:6]: Also, should we bind to `~6.0.0` to only get bug fixes, or, are you happy with potentially getting features also?

Comment (abul-wcse) [reply composer.json:6]: Nope as meteor is a self container package

Comment (jedkirby) [reply composer.json:6]: 👍

File: src/Package/Provider/BasicHttp/BasicHttpPackageProvider.php (hunk 4 above)
Comment (abul-wcse) [reply src/Package/Provider/BasicHttp/BasicHttpPackageProvider.php]: 👍

**Comment by jedkirby** [comment]
@abul-wcse - I know this is unrelated to your work, but, we need to fix `phpunit/phpunit` and `mockery/mockery` to specific versions within the `composer.json` file now as they've both released new stable versions which are incompatible with some PHP versions. You can see the tests are failing on this PR, which is related to that issue.

File: src/Package/Provider/BasicHttp/BasicHttpPackageProvider.php (hunk 4 above)
Comment (abul-wcse) [reply src/Package/Provider/BasicHttp/BasicHttpPackageProvider.php]: Done

File: src/Package/Provider/BasicHttp/BasicHttpPackageProvider.php (hunk 5 above)
Comment (abul-wcse) [reply src/Package/Provider/BasicHttp/BasicHttpPackageProvider.php]: Nope

File: src/Package/Provider/BasicHttp/BasicHttpPackageProvider.php (hunk 3 above)
Comment (abul-wcse) [reply src/Package/Provider/BasicHttp/BasicHttpPackageProvider.php]: Done

**Comment by abul-wcse** [comment]
I have updated the version of mockery to suite the phpUnit version we use

File: composer.json (hunk 1 above)
Comment (abul-wcse) [reply composer.json:6]: Done

**Review by jedkirby — APPROVED** [review]
👍

**Pull request merged**